### PR TITLE
implement optional stages in toast.promise

### LINKED ISSUE #644: Make the stages of toast.promise optional
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
`toast.promise` expects all three stages to be defined in `ToastPromiseParams` e.g. `success`, `pending` and `error`. 

**What is the expected behavior?**
As a developer, I want each of these properties to be optional so that I can utilise the `toast.promise` function for resolving promises but have more control over what toasts are shown.

For example, I may want to resolve a promise whereby only an error toast is required and so I would like to write:

```javascript
toast.promise(somePromiseFn, { error: 'Something went wrong!' });
```

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
N/A

### PR DESCRIPTION
**Pull request by wilson-webdev**
- Make the three props in the `toast.promise` object optional to enable greater flexibility
- Add checkbox to the example to disable the pending state

Resolves #644

### CODE REVIEW

File: example/components/constants.ts [hunk 1]
```diff
@@ -32,6 +32,10 @@ export const flags = [
   {
     id: 'draggable',
     label: 'Allow to drag and close the toast'
+  },
+  {
+    id: 'disablePendingPromise',
+    label: 'Disable pending stage on Promise toast'
```
Comment (fkhadra) [inline example/components/constants.ts:38]: I don't think this feature require a new prop

File: src/core/toast.tsx [hunk 2]
```diff
@@ -121,23 +121,50 @@ toast.loading = (content: ToastContent, options?: ToastOptions) =>
...
+      isLoading,
+      ...toastParams.options
+    });
+  };
+
+  // Initialise the id before calling the promise if a pending state has been provided
+  if (pending) {
+    initialiseToast(pending, true);
+  }
+
```
Comment (fkhadra) [inline src/core/toast.tsx:167]: I think that the implementation can be simplified as follow
```tsx
interface ToastPromiseParams {
  pending?: string | UpdateOptions;
  success?: string | UpdateOptions;
  error?: string | UpdateOptions;
}

function handlePromise<T>(
  promise: Promise<T> | (() => Promise<T>),
  { pending, error, success }: ToastPromiseParams,
  options?: ToastOptions
) {
  let id: Id;
 
  if (pending) {
    id = isStr(pending)
      ? toast.loading(pending, options)
      : toast.loading(pending.render, {
          ...options,
          ...(pending as ToastOptions)
        });
  }

  const resetParams = {
    isLoading: null,
    autoClose: null,
    closeOnClick: null,
    closeButton: null,
    draggable: null
  };

  const resolver = (
    type: TypeOptions,
    input: string | UpdateOptions,
    result: T
  ) => {
    const baseParams = {
      type,
      ...resetParams,
      ...options,
      data: result
    };
    const params = isStr(input) ? { render: input } : input;
    
   // if the id is set we know that it's an update
    if (id) {
      toast.update(id, {
        ...baseParams,
        ...params
      });
    } else {
     // using toast.promise without loading
      toast(params.render, {
        ...baseParams,
        ...params
      } as ToastOptions);
    }

    return result;
  };
  const p = isFn(promise) ? promise() : promise;

  //call the resolvers only when needed
  p.then(result => success && resolver('success', success, result)).catch(
    err => error && resolver('error', error, err)
  );

  return p;
}
```

**Review by fkhadra — COMMENTED** [review]
Hey thanks for the PR, really appreciate that. Adding a new prop for this use case is not needed, I prefer to avoid extending the API if we can.

When it comes to the implementation itself, I think this can be simplified as well (this is subjective 😆). I've added mine in the review, let me know what you think about it.  

I've created a patch if you want to test it. https://drive.google.com/file/d/1q5DDI7svgl_X-mePxr53n3cDuYKHmRa6/view?usp=sharing

**Comment by wilson-webdev** [comment]
@fkhadra I've pushed the changes suggested, along with removing the edits to the example page. I agree it's nice and simple!

One thing I have noticed is that if you provide a `pending` stage and not `success`/`error`, it hangs on the pending state. We could add a bit more logic to the promise then/catch chain to check what's defined, and then remove the toast if appropriate. Up to you! 😃

**Comment by wilson-webdev** [comment]
@fkhadra any update on this?

**Pull request merged**